File: Reversi/Reversi/Reversi.cpp
//======================================
//	リバーシ
//======================================
#include "Reversi.h"
#include "Vector2.h"
#include "Utility.h"
#include <stdio.h>  // printf()
#include <assert.h> // assert()
// 関数プロトタイプ
bool inBoard(int x, int y);
const char* getTurnName(Cell turn);

const char* diskAA[] = {
	"・",  // NONE
	"●",  // BLACK
	"〇",  // WHITE
	"■",  // OUT
};

// ボードの指定位置に代入
void SetBoard(Reversi* reversi, int x, int y, Cell value)
{
	//
	// ★ここをコーディングしてください。
	// inBoard()を確認して　boardの(x,y)座標にvalueを格納します
	// inBoard()外なら、何もしません
	//
	if (inBoard(x, y))
		reversi->board[y][x] = value;
}
// ボードの指定位置を取得
Cell GetBoard(Reversi* reversi, int x, int y)
{
	//
	// ★ここをコーディングしてください。
	// inBoard()を確認して　boardの(x,y)座標の値を返します。
	// inBoard()外なら、OUT を返してください。
	//
	if (inBoard(x, y))
		return reversi->board[y][x];
	return OUT;
}
// 初期化
void Init(Reversi* reversi, Mode mode)
{
	reversi->mode = mode;
	//
	// ★ここをコーディングしてください。
	// ボードを初期状態にします。
	// ・・・・・・・・
	// ・・・・・・・・
	// ・・・・・・・・
	// ・・・〇●・・・
	// ・・・●〇・・・
	// ・・・・・・・・
	// ・・・・・・・・
	// ・・・・・・・・
	//
	for (int y = 0; y < BOARD_HEI; y++) {
		for (int x = 0; x < BOARD_WID; x++) {
			SetBoard(reversi, x, y, NONE);
		}
	}
	SetBoard(reversi, 3, 4, BLACK);
	SetBoard(reversi, 4, 3, BLACK);
	SetBoard(reversi, 3, 3, WHITE);
	SetBoard(reversi, 4, 4, WHITE);
	reversi->turn = BLACK;
}
// 次のターンへ
void NextTurn(Reversi* reversi)
{
	reversi->turn = GetOpponent(reversi->turn);
}
// 指定位置に石がおけるか
bool CheckCanPlace(Reversi* reversi, Cell myself, Vector2 pos, bool turnOver)
{
	bool canPlace = false;
	// 既に置かれている?
	if (GetBoard(reversi, pos.x, pos.y) != NONE) {
		return false;
	}
	Cell opponent = GetOpponent(myself);
	// ８方向について、
	for (int d = 0; d < DIR_MAX; d++) {
		Vector2 scan = pos;
		Vector2 dir = GetDirVector2((DIRECTION)d);
		// 隣は相手コマ?
		
		// ★ここをコーディングしてください。
		// scanをdir進めて、
		// (scan.x,scan.y)のボードが相手コマでなければ
		// 置けないので 次のfor()へ移行します(continue)
		AddVector2(&scan, &dir);
		if (GetBoard(reversi, scan.x, scan.y) != opponent)
			continue;
		// 相手コマでなくなるまで進める
		Cell tmp;
		//
		// ★ここをコーディングしてください。
		// scanをdir進めていって
		// (scan.x,scan.y)のボードが相手コマでなくなるまで
		// 繰り返します
		do {
			AddVector2(&scan, &dir);
			tmp = GetBoard(reversi, scan.x, scan.y);
		} while (tmp == opponent);
		// 自分のコマなら、place可
		if (tmp == myself) {
			canPlace = true;
			// ひっくりかえす
			if (turnOver) {
				scan = pos;
				do {
					AddVector2(&scan, &dir);
					tmp = GetBoard(reversi, scan.x, scan.y);
					if (tmp == opponent) {
						SetBoard(reversi, scan.x, scan.y, myself);
					}
				} while (tmp == opponent);
			}
		}
		// 「置ける」turnOverしないなら、終了
		if (canPlace && turnOver == false) {
			break;
		}
	}
	return canPlace;
}
// コマの数を数える
int getDiskCount(Reversi* reversi, Cell myself)
{
	int count = 0;
	//
	// ★ここをコーディングしてください
	// ボードを走査して、myself と一致するものをカウントします。
	//
	for (int y = 0; y < BOARD_HEI; y++) {
		for (int x = 0; x < BOARD_WID; x++) {
			if (GetBoard(reversi, x, y) == myself)
				count++;
		}
	}
	return count;
}
// 石がどこかに置けるか?
bool CheckCanPlaceAll(Reversi* reversi, Cell myself)
{
	for (int y = 0; y < BOARD_HEI; y++) {
		for (int x = 0; x < BOARD_WID; x++) {
			Vector2 pos = { x,y };
			if (CheckCanPlace(reversi, myself, pos)) {
				return true;
			}
		}
	}
	return false;
}
// 置ける場所をリスティング
void ListCanPlaceAll(Reversi* reversi, Cell myself, Vector2List* list)
{
	Vector2 pos;
	for (pos.y = 0; pos.y < BOARD_HEI; pos.y++) {
		for (pos.x = 0; pos.x < BOARD_WID; pos.x++) {
			if (CheckCanPlace(reversi, myself, pos)) {
				AddVector2List(list, pos);
			}
		}
	}
}
// 画面描画
void DrawScreen(Reversi* reversi, Vector2 pos, DrawStatus stat)
{
	ClearScreen();
	bool isCursor = (stat == IN_PLAY) && IsHumanPlayer(reversi);
	//
	// ★ここをコーディングしてください。
	// ボードを走査してcellによって diskAA[] を表示します。
	// isCorsorが真なら、posに対応した所、右に←を下に↑を表示します。
	// ・・・・・・・・・
	// ・・・・・・・・・
	// ・・・・・・・・・
	// ・・・〇●・・・・←
	// ・・・●〇・・・・
	// ・・・・・・・・・
	// ・・・・・・・・・
	// ・・・・・・・・・
	// 　　　↑
	//
	for (int y = 0; y < BOARD_HEI; y++) {
		for (int x = 0; x < BOARD_WID; x++) {
			Cell tmp = GetBoard(reversi, x, y);
			printf("%s", diskAA[tmp]);
		}
		if (isCursor && pos.y == y) {
			printf("←");
		}
		putchar('\n');
	}
	for (int x = 0; x < BOARD_WID; x++) {
		if (isCursor && pos.x == x)
			printf("↑");
		else
			printf("　");
	}
	putchar('\n');

	//現在のコマ数表示
	printf("現在のコマ数：黒%d,白%d\n", getDiskCount(reversi, BLACK), getDiskCount(reversi, WHITE));

	switch (stat) {
	case IN_PLAY:
	case NO_PLACE:
		// 
		// ★ここをコーディングしてください。
		// 「～のターンです」を表示します
		// 更に NO_PLACEのときは
		// 「打てるところがありません」を表示します
		//
		printf("%sのターンです\n", getTurnName(reversi->turn));
		if (stat == NO_PLACE)
			printf("打てるところがありません\n");
		break;
	case RESULT:
	{
		int blackCount = getDiskCount(reversi, BLACK);
		int whiteCount = getDiskCount(reversi, WHITE);
		// 
		// ★ここをコーディングしてください
		// 「黒xx-白xx」(xxはコマ数)を表示します
		// コマ数が同じなら
		// 「引き分けです」を表示します
		// そうでなければ
		// 「～の勝ちです」を表示します
		// 
		printf("%s%d-%s%d\n",
			getTurnName(BLACK), blackCount,
			getTurnName(WHITE), whiteCount
		);
		if (blackCount == whiteCount)
			puts("引き分けです");
		else {
			Cell winner = (blackCount > whiteCount) ? BLACK : WHITE;
			printf("%sの勝ちです。\n", getTurnName(winner));
		}
	}
	break;
	}
}

bool IsHumanPlayer(Reversi* reversi)
{
	switch (reversi->mode) {
	case MODE_1P:
		return reversi->turn == BLACK;
	case MODE_2P:
		return true;
	default:
	case MODE_WATCH:
		return false;
	}
}

const char* getTurnName(Cell turn)
{
	if (turn == BLACK) {
		return "黒";
	}
	if (turn == WHITE) {
		return "白";
	}
	assert(false);
	return "";
}
// 相手の色を取得
Cell GetOpponent(Cell value)
{
	if (value == WHITE) {
		return BLACK;
	}
	if (value == BLACK) {
		return WHITE;
	}
	assert(false);
	return NONE;
}
// 座標がボード内か?
bool inBoard(int x, int y)
{
	return 0 <= x && x < BOARD_WID
		&& 0 <= y && y < BOARD_HEI;
}

File: Reversi_CS/Reversi_CS/Reversi.cs
//======================================
//      リバーシ　
//======================================
using System.Diagnostics;  // Debug
using System.Collections.Generic; // List<T>
using Vector2 = GP2.Vector2;
using Direction8 = GP2.Direction8;
using Utility = GP2.Utility;

namespace Reversi_CS
{
    // ゲームモード
    enum GameMode
    {
        _1P,
        _2P,
        _Watch,
    }
    // ボードのセル
    enum Cell
    {
        None,   // 空白
        Black,  // 黒のコマ
        White,  // 白のコマ
        Out,    // 盤の外
    }
    // 描画ステート
    enum DrawStatus
    {
        InPlay,   // プレイ中
        NoSpace,  // 打てるところがない
        Result,   // リザルト
    }

    internal class Reversi
    {
        public const int BOARD_WID = 8;
        public const int BOARD_HEI = 8;
        protected Cell[,] m_board;
        protected Cell m_turn;
        protected GameMode m_mode;

        // コンストラクター
        public Reversi(GameMode mode)
        {
            m_mode = mode;

            m_board = new Cell[BOARD_HEI, BOARD_WID];
            // 盤を初期化
            for (int y = 0; y < BOARD_HEI; y++)
            {
                for (int x = 0; x < BOARD_WID; x++)
                {
                    SetBoard(x, y, Cell.None);
                }
            }
            SetBoard(3, 4, Cell.Black);
            SetBoard(4, 3, Cell.Black);
            SetBoard(3, 3, Cell.White);
            SetBoard(4, 4, Cell.White);
            m_turn = Cell.Black;
        }
        // プロパティ turn
        public Cell turn
        {
            get { return m_turn; }
        }
        // 次のターンへ
        public void NextTurn()
        {
            m_turn = GetOpponent(m_turn);
        }
        // 指定位置に置けるか?
        public bool CheckCanPlace(Cell myself, Vector2 pos, bool turnover)
        {
            bool canPlace = false;
            // 既に置かれている?
            if (GetBoard(pos.x, pos.y) != Cell.None)
            {
                return false;
            }

            Cell opponent = GetOpponent(myself);
            // ８方向について、
            for (int d = 0; d < (int)Direction8.MAX; d++)
            {
                Vector2 scan = pos;
                Vector2 dir = Vector2.GetDirVector2((Direction8)d);
                // 隣は相手コマ?
                scan += dir;
                if (GetBoard(scan.x, scan.y) != opponent)
                {
                    continue;
                }
                Cell tmp;
                do
                {
                    scan += dir;
                    tmp = GetBoard(scan.x, scan.y);
                } while (tmp == opponent);
                // 自分のコマなら、place可
                if (tmp == myself)
                {
                    canPlace = true;
                    if (turnover)
                    {
                        scan = pos;
                        do
                        {
                            scan += dir;
                            tmp = GetBoard(scan.x, scan.y);
                            if (tmp == opponent)
                            {
                                SetBoard(scan.x, scan.y, myself);
                            }
                        } while (tmp == opponent);
                    }
                }
                if (canPlace && turnover == false)
                {
                    break;
                }
            }
            return canPlace;
        }
        // 盤上のコマ数を数える
        int getDiskCount(Cell myself)
        {
            int count = 0;
            for (int y = 0; y < BOARD_HEI; y++)
            {
                for (int x = 0; x < BOARD_WID; x++)
                {
                    if (GetBoard(x, y) == myself)
                    {
                        count++;
                    }
                }
            }
            return count;
        }
        public bool CheckCanPlaceAll(Cell myself)
        {
            Vector2 pos = new Vector2(0, 0);
            for (int y = 0; y < BOARD_HEI; y++)
            {
                for (int x = 0; x < BOARD_WID; x++)
                {
                    pos.Set(x, y);
                    if (CheckCanPlace(myself, pos, false))
                    {
                        return true;
                    }
                }
            }
            return false;
        }
        // 打てるところをリスティング
        public void ListCanPlaceAll(Cell myself, List<Vector2> list)
        {
            Vector2 pos = new Vector2(0, 0);
            for (int y = 0; y < BOARD_HEI; y++)
            {
                for (int x = 0; x < BOARD_WID; x++)
                {
                    pos.Set(x, y);
                    if (CheckCanPlace(myself, pos, false))
                    {
                        list.Add(pos);
                    }
                }
            }
        }

        protected static string[] diskAA = new string[]
        {
                "・",  // NONE
                "●",  // BLACK
                "〇",  // WHITE
                "■",  // OUT
        };

        // 画面描画
        public void DrawScreen(Vector2 pos, DrawStatus stat)
        {
            Utility.ClearScreen();
            bool isCursor = (stat == DrawStatus.InPlay) && IsHumanPlayer();
            for (int y = 0; y < BOARD_HEI; y++)
            {
                for (int x = 0; x < BOARD_WID; x++)
                {
                    Cell tmp = GetBoard(x, y);
                    Utility.Printf("{0}", diskAA[(int)tmp]);
                }
                if (isCursor && pos.y == y)
                {
                    Utility.Printf("←");
                }
                Utility.Putchar('\n');
            }
            for (int x = 0; x < BOARD_WID; x++)
            {
                if (isCursor && pos.x == x)
                {
                    Utility.Printf("↑");
                }
                else
                {
                    Utility.Printf("　");
                }
            }
            Utility.Putchar('\n');

            switch (stat)
            {
                case DrawStatus.InPlay:
                case DrawStatus.NoSpace:
                    Utility.Printf("{0}のターンです\n", getTurnName(m_turn));
                    if (stat == DrawStatus.NoSpace)
                    {
                        Utility.Printf("打てるところがありません\n");
                    }
                    break;
                case DrawStatus.Result:
                    {
                        int blackCount = getDiskCount(Cell.Black);
                        int whiteCount = getDiskCount(Cell.White);
                        Utility.Printf("{0}{1}-{2}{3}\n\n"
                            , getTurnName(Cell.Black), blackCount
                            , getTurnName(Cell.White), whiteCount
                            );
                        if (blackCount == whiteCount)
                        {
                            Utility.Puts("引き分けです");
                        }
                        else
                        {
                            Cell winner = blackCount >= whiteCount ? Cell.Black : Cell.White;
                            Utility.Printf("{0}の勝ちです\n", getTurnName(winner));
                        }
                    }
                    break;
            }
            Utility.PrintOut();
        }
        // 人間プレーヤのターンか?
        public bool IsHumanPlayer()
        {
            switch (m_mode)
            {
                case GameMode._1P:
                    return m_turn == Cell.Black;
                case GameMode._2P:
                    return true;
                case GameMode._Watch:
                    return false;
            }
            Debug.Assert(false);
            return false;
        }

        // 相手を取得
        public static Cell GetOpponent(Cell cell)
        {

            switch (cell)
            {
                case Cell.Black: return Cell.White;
                case Cell.White: return Cell.Black;
            }
            Debug.Assert(false);
            return Cell.None;
        }

        // ボードのセッター
        public void SetBoard(int x, int y, Cell value)
        {
            if (inBoard(x, y))
            {
                m_board[y, x] = value;
            }
        }
        // ボードのゲッター
        public Cell GetBoard(int x, int y)
        {
            if (inBoard(x, y))
            {
                return m_board[y, x];
            }
            return Cell.Out;
        }
        protected static string getTurnName(Cell turn)
        {
            switch (turn)
            {
                case Cell.Black: return "黒";
                case Cell.White: return "白";
            }
            Debug.Assert(false);
            return "";
        }

        // ボード内か?
        protected static bool inBoard(int x, int y)
        {
            return 0 <= x && x < BOARD_WID
                && 0 <= y && y < BOARD_HEI;
        }
    }
}